// gui2-fw-lib/esm2015/lib/remote/urlfn.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
/*
 * Copyright 2018-present Open Networking Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Injectable, Inject } from '@angular/core';
import { LogService } from '../log.service';
import * as i0 from "@angular/core";
import * as i1 from "../log.service";
/** @type {?} */
const UICONTEXT = '/onos/ui/';
/** @type {?} */
const RSSUFFIX = UICONTEXT + 'rs/';
/** @type {?} */
const WSSUFFIX = UICONTEXT + 'websock/';
/**
 * ONOS GUI -- Remote -- General Purpose URL Functions
 */
export class UrlFnService {
    /**
     * @param {?} log
     * @param {?} w
     */
    constructor(log, w) {
        this.log = log;
        this.w = w;
        this.log.debug('UrlFnService constructed');
    }
    /**
     * @param {?} protocol
     * @return {?}
     */
    matchSecure(protocol) {
        /** @type {?} */
        const p = this.w.location.protocol;
        /** @type {?} */
        const secure = (p === 'https' || p === 'wss');
        return secure ? protocol + 's' : protocol;
    }
    /* A little bit of funky here. It is possible that ONOS sits
         * behind a proxy and has an app prefix, e.g.
         *      http://host:port/my/app/onos/ui...
         * This bit of regex grabs everything after the host:port and
         * before the UICONTEXT (/onos/ui/) and uses that as an app
         * prefix by inserting it back into the WS URL.
         * If no prefix, then no insert.
         */
    /**
     * @param {?} protocol
     * @param {?=} port
     * @param {?=} host
     * @return {?}
     */
    urlBase(protocol, port = '', host = '') {
        /** @type {?} */
        const match = this.w.location.href.match('.*//[^/]+/(.+)' + UICONTEXT);
        /** @type {?} */
        const appPrefix = match ? '/' + match[1] : '';
        return this.matchSecure(protocol) +
            '://' +
            (host === '' ? this.w.location.hostname : host) +
            ':' +
            (port === '' ? this.w.location.port : port) +
            appPrefix;
    }
    /**
     * @param {?} suffix
     * @return {?}
     */
    httpPrefix(suffix) {
        return this.urlBase('http') + suffix;
    }
    /**
     * @param {?} suffix
     * @param {?} wsport
     * @param {?} host
     * @return {?}
     */
    wsPrefix(suffix, wsport, host) {
        return this.urlBase('ws', wsport, host) + suffix;
    }
    /**
     * @param {?} path
     * @return {?}
     */
    rsUrl(path) {
        return this.httpPrefix(RSSUFFIX) + path;
    }
    /**
     * @param {?} path
     * @param {?=} wsport
     * @param {?=} host
     * @return {?}
     */
    wsUrl(path, wsport, host) {
        return this.wsPrefix(WSSUFFIX, wsport, host) + path;
    }
}
UrlFnService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] },
];
/** @nocollapse */
UrlFnService.ctorParameters = () => [
    { type: LogService },
    { type: undefined, decorators: [{ type: Inject, args: ['Window',] }] }
];
/** @nocollapse */ UrlFnService.ngInjectableDef = i0.defineInjectable({ factory: function UrlFnService_Factory() { return new UrlFnService(i0.inject(i1.LogService), i0.inject("Window")); }, token: UrlFnService, providedIn: "root" });
if (false) {
    /** @type {?} */
    UrlFnService.prototype.log;
    /** @type {?} */
    UrlFnService.prototype.w;
}
