// gui2-fw-lib/esm5/lib/widget/detailspanel.base.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { PanelBaseImpl } from './panel.base';
import { Output, EventEmitter, Input } from '@angular/core';
/**
 * A generic model of the data returned from the *DetailsResponse
 * @record
 */
export function DetailsResponse() { }
if (false) {
    /** @type {?} */
    DetailsResponse.prototype.details;
}
/**
 * Extends the PanelBase abstract class specifically for showing details
 *
 * This makes another call through WSS to the server for specific
 * details to fill the panel with
 *
 * This replaces the detailspanel service in the old gui
 * @abstract
 */
var DetailsPanelBaseImpl = /** @class */ (function (_super) {
    tslib_1.__extends(DetailsPanelBaseImpl, _super);
    function DetailsPanelBaseImpl(fs, ls, log, wss, tag) {
        var _this = _super.call(this, fs, ls, log) || this;
        _this.fs = fs;
        _this.ls = ls;
        _this.log = log;
        _this.wss = wss;
        _this.tag = tag;
        _this.closeEvent = new EventEmitter();
        _this.handlers = [];
        _this.detailsData = {};
        _this.closed = false;
        _this.root = tag + 's';
        _this.req = tag + 'DetailsRequest';
        _this.resp = tag + 'DetailsResponse';
        return _this;
    }
    /**
     * When the details panel is created set up a listener on
     * Web Socket for details responses
     */
    /**
     * When the details panel is created set up a listener on
     * Web Socket for details responses
     * @return {?}
     */
    DetailsPanelBaseImpl.prototype.init = /**
     * When the details panel is created set up a listener on
     * Web Socket for details responses
     * @return {?}
     */
    function () {
        var _this = this;
        this.wss.bindHandlers(new Map([
            [this.resp, function (data) { return _this.detailsPanelResponseCb(data); }]
        ]));
        this.handlers.push(this.resp);
    };
    /**
     * When the details panel is destroyed this should be called to
     * de-register from the WebSocket
     */
    /**
     * When the details panel is destroyed this should be called to
     * de-register from the WebSocket
     * @return {?}
     */
    DetailsPanelBaseImpl.prototype.destroy = /**
     * When the details panel is destroyed this should be called to
     * de-register from the WebSocket
     * @return {?}
     */
    function () {
        this.wss.unbindHandlers(this.handlers);
    };
    /**
     * A callback that executes when the details data that was requested
     * on WebSocketService arrives.
     */
    /**
     * A callback that executes when the details data that was requested
     * on WebSocketService arrives.
     * @param {?} data
     * @return {?}
     */
    DetailsPanelBaseImpl.prototype.detailsPanelResponseCb = /**
     * A callback that executes when the details data that was requested
     * on WebSocketService arrives.
     * @param {?} data
     * @return {?}
     */
    function (data) {
        this.detailsData = data['details'];
    };
    /**
     * Details Panel Data Request - should be called whenever row id changes
     */
    /**
     * Details Panel Data Request - should be called whenever row id changes
     * @param {?} query
     * @return {?}
     */
    DetailsPanelBaseImpl.prototype.requestDetailsPanelData = /**
     * Details Panel Data Request - should be called whenever row id changes
     * @param {?} query
     * @return {?}
     */
    function (query) {
        this.closed = false;
        // Do not send if the Web Socket hasn't opened
        if (this.wss.isConnected()) {
            if (this.fs.debugOn('panel')) {
                this.log.debug('Details panel data REQUEST:', this.req, query);
            }
            this.wss.sendEvent(this.req, query);
        }
    };
    /**
     * this should be called when the details panel close button is clicked
     */
    /**
     * this should be called when the details panel close button is clicked
     * @return {?}
     */
    DetailsPanelBaseImpl.prototype.close = /**
     * this should be called when the details panel close button is clicked
     * @return {?}
     */
    function () {
        this.closed = true;
        this.id = null;
        this.closeEvent.emit(this.id);
    };
    DetailsPanelBaseImpl.propDecorators = {
        id: [{ type: Input }],
        closeEvent: [{ type: Output }]
    };
    return DetailsPanelBaseImpl;
}(PanelBaseImpl));
export { DetailsPanelBaseImpl };
if (false) {
    /** @type {?} */
    DetailsPanelBaseImpl.prototype.id;
    /** @type {?} */
    DetailsPanelBaseImpl.prototype.closeEvent;
    /** @type {?} */
    DetailsPanelBaseImpl.prototype.root;
    /** @type {?} */
    DetailsPanelBaseImpl.prototype.req;
    /** @type {?} */
    DetailsPanelBaseImpl.prototype.resp;
    /** @type {?} */
    DetailsPanelBaseImpl.prototype.handlers;
    /** @type {?} */
    DetailsPanelBaseImpl.prototype.detailsData;
    /** @type {?} */
    DetailsPanelBaseImpl.prototype.closed;
    /** @type {?} */
    DetailsPanelBaseImpl.prototype.fs;
    /** @type {?} */
    DetailsPanelBaseImpl.prototype.ls;
    /** @type {?} */
    DetailsPanelBaseImpl.prototype.log;
    /** @type {?} */
    DetailsPanelBaseImpl.prototype.wss;
    /** @type {?} */
    DetailsPanelBaseImpl.prototype.tag;
}
