// gui2-fw-lib/esm5/lib/widget/table.base.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/** @type {?} */
var REFRESH_INTERVAL = 2000;
/** @type {?} */
var SEARCH_REGEX = '\\W';
/**
 * Model of table annotations within this table base class
 * @record
 */
export function TableAnnots() { }
if (false) {
    /** @type {?} */
    TableAnnots.prototype.noRowsMsg;
}
/**
 * A model of data returned from Web Socket in a TableResponse
 *
 * There is an interface extending from this one in the parent component
 * @record
 */
export function TableResponse() { }
if (false) {
    /** @type {?} */
    TableResponse.prototype.annots;
}
/**
 * A criteria for filtering the tableData
 * @record
 */
export function TableFilter() { }
if (false) {
    /** @type {?} */
    TableFilter.prototype.queryStr;
    /** @type {?} */
    TableFilter.prototype.queryBy;
    /** @type {?} */
    TableFilter.prototype.sortBy;
}
/** @enum {string} */
var SortDir = {
    asc: 'asc', desc: 'desc',
};
export { SortDir };
/**
 * A structure to format sort params for table
 * This is sent to WebSocket as part of table request
 * @record
 */
export function SortParams() { }
if (false) {
    /** @type {?} */
    SortParams.prototype.firstCol;
    /** @type {?} */
    SortParams.prototype.firstDir;
    /** @type {?} */
    SortParams.prototype.secondCol;
    /** @type {?} */
    SortParams.prototype.secondDir;
}
/**
 * @record
 */
export function PayloadParams() { }
if (false) {
    /** @type {?} */
    PayloadParams.prototype.devId;
}
/**
 * ONOS GUI -- Widget -- Table Base class
 * @abstract
 */
var /**
 * ONOS GUI -- Widget -- Table Base class
 * @abstract
 */
TableBaseImpl = /** @class */ (function () {
    function TableBaseImpl(fs, ls, log, wss, tag, idKey, selCb // Function
    ) {
        if (idKey === void 0) { idKey = 'id'; }
        if (selCb === void 0) { selCb = function () { return ({}); }; }
        var _this = this;
        this.fs = fs;
        this.ls = ls;
        this.log = log;
        this.wss = wss;
        this.tag = tag;
        this.idKey = idKey;
        this.selCb = selCb;
        this.changedData = [];
        // Function
        this.parentSelCb = null;
        // Function
        this.selId = undefined;
        this.tableData = [];
        // Function
        this.autoRefresh = true;
        this.autoRefreshTip = 'Toggle auto refresh'; // TODO: get LION string
        this.refreshPromise = null;
        this.handlers = [];
        this.root = tag + 's';
        this.req = tag + 'DataRequest';
        this.resp = tag + 'DataResponse';
        this.selectCallback = this.rowSelectionCb;
        this.toggleRefresh = function () {
            _this.autoRefresh = !_this.autoRefresh;
            _this.autoRefresh ? _this.startRefresh() : _this.stopRefresh();
        };
        // Mapped to the search and searchBy inputs in template
        // Changes are handled through TableFilterPipe
        this.tableDataFilter = (/** @type {?} */ ({
            queryStr: '',
            queryBy: '$',
        }));
    }
    /**
     * @return {?}
     */
    TableBaseImpl.prototype.init = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.wss.bindHandlers(new Map([
            [this.resp, function (data) { return _this.tableDataResponseCb(data); }]
        ]));
        this.handlers.push(this.resp);
        this.annots = (/** @type {?} */ ({
            noRowsMsg: ''
        }));
        // Now send the WebSocket request and make it repeat every 2 seconds
        this.requestTableData();
        this.startRefresh();
        this.log.debug('TableBase initialized. Calling ', this.req, 'every', REFRESH_INTERVAL, 'ms');
    };
    /**
     * @return {?}
     */
    TableBaseImpl.prototype.destroy = /**
     * @return {?}
     */
    function () {
        this.wss.unbindHandlers(this.handlers);
        this.stopRefresh();
        this.ls.stop();
    };
    /**
     * A callback that executes when the table data that was requested
     * on WebSocketService arrives.
     *
     * Happens every 2 seconds
     */
    /**
     * A callback that executes when the table data that was requested
     * on WebSocketService arrives.
     *
     * Happens every 2 seconds
     * @param {?} data
     * @return {?}
     */
    TableBaseImpl.prototype.tableDataResponseCb = /**
     * A callback that executes when the table data that was requested
     * on WebSocketService arrives.
     *
     * Happens every 2 seconds
     * @param {?} data
     * @return {?}
     */
    function (data) {
        this.ls.stop();
        /** @type {?} */
        var newTableData = Array.from(data[this.root]);
        this.annots.noRowsMsg = data.annots.no_rows_msg;
        // If the parents onResp() function is set then call it
        if (this.responseCallback) {
            this.responseCallback(data);
        }
        this.changedData = [];
        // checks if data changed for row flashing
        if (JSON.stringify(newTableData) !== JSON.stringify(this.tableData)) {
            this.log.debug('table data has changed');
            /** @type {?} */
            var oldTableData = this.tableData;
            this.tableData = tslib_1.__spread(newTableData); // ES6 spread syntax
            // only flash the row if the data already exists
            if (oldTableData.length > 0) {
                for (var idx in newTableData) {
                    if (!this.fs.containsObj(oldTableData, newTableData[idx])) {
                        this.changedData.push(newTableData[idx][this.idKey]);
                    }
                }
            }
        }
    };
    /**
     * Table Data Request
     * Pass in sort parameters and the set will be returned sorted
     * In the old GUI there was also a query parameter, but this was not
     * implemented on the server end
     */
    /**
     * Table Data Request
     * Pass in sort parameters and the set will be returned sorted
     * In the old GUI there was also a query parameter, but this was not
     * implemented on the server end
     * @return {?}
     */
    TableBaseImpl.prototype.requestTableData = /**
     * Table Data Request
     * Pass in sort parameters and the set will be returned sorted
     * In the old GUI there was also a query parameter, but this was not
     * implemented on the server end
     * @return {?}
     */
    function () {
        /** @type {?} */
        var p = Object.assign({}, this.sortParams, this.payloadParams);
        // Allow it to sit in pending events
        if (this.wss.isConnected()) {
            if (this.fs.debugOn('table')) {
                this.log.debug('Table data REQUEST:', this.req, p);
            }
            this.wss.sendEvent(this.req, p);
            this.ls.start();
        }
    };
    /**
     * Row Selected
     */
    /**
     * Row Selected
     * @param {?} event
     * @param {?} selRow
     * @return {?}
     */
    TableBaseImpl.prototype.rowSelectionCb = /**
     * Row Selected
     * @param {?} event
     * @param {?} selRow
     * @return {?}
     */
    function (event, selRow) {
        /** @type {?} */
        var selId = selRow[this.idKey];
        this.selId = (this.selId === selId) ? undefined : selId;
        this.log.debug('Row', selId, 'selected');
        if (this.parentSelCb) {
            this.parentSelCb(event, selRow);
        }
    };
    /**
     * autoRefresh functions
     */
    /**
     * autoRefresh functions
     * @return {?}
     */
    TableBaseImpl.prototype.startRefresh = /**
     * autoRefresh functions
     * @return {?}
     */
    function () {
        var _this = this;
        this.refreshPromise =
            setInterval(function () {
                if (!_this.ls.waiting()) {
                    if (_this.fs.debugOn('table')) {
                        _this.log.debug('Refreshing ' + _this.root + ' page');
                    }
                    _this.requestTableData();
                }
            }, REFRESH_INTERVAL);
    };
    /**
     * @return {?}
     */
    TableBaseImpl.prototype.stopRefresh = /**
     * @return {?}
     */
    function () {
        if (this.refreshPromise) {
            clearInterval(this.refreshPromise);
            this.refreshPromise = null;
        }
    };
    /**
     * @param {?} id
     * @return {?}
     */
    TableBaseImpl.prototype.isChanged = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        return (this.fs.inArray(id, this.changedData) === -1) ? false : true;
    };
    /**
     * A dummy implementation of the lionFn until the response is received and the LION
     * bundle is received from the WebSocket
     */
    /**
     * A dummy implementation of the lionFn until the response is received and the LION
     * bundle is received from the WebSocket
     * @param {?} key
     * @return {?}
     */
    TableBaseImpl.prototype.dummyLion = /**
     * A dummy implementation of the lionFn until the response is received and the LION
     * bundle is received from the WebSocket
     * @param {?} key
     * @return {?}
     */
    function (key) {
        return '%' + key + '%';
    };
    /**
     * Change the sort order of the data returned
     *
     * sortParams are passed to the server by WebSocket and the data is
     * returned sorted
     *
     * This is usually assigned to the (click) event on a column, and the column
     * name passed in e.g. (click)="onSort('origin')
     * If the column that is passed in is already the firstCol, then reverse its direction
     * If a new column is passed in, then make the existing col the 2nd sort order
     */
    /**
     * Change the sort order of the data returned
     *
     * sortParams are passed to the server by WebSocket and the data is
     * returned sorted
     *
     * This is usually assigned to the (click) event on a column, and the column
     * name passed in e.g. (click)="onSort('origin')
     * If the column that is passed in is already the firstCol, then reverse its direction
     * If a new column is passed in, then make the existing col the 2nd sort order
     * @param {?} colName
     * @return {?}
     */
    TableBaseImpl.prototype.onSort = /**
     * Change the sort order of the data returned
     *
     * sortParams are passed to the server by WebSocket and the data is
     * returned sorted
     *
     * This is usually assigned to the (click) event on a column, and the column
     * name passed in e.g. (click)="onSort('origin')
     * If the column that is passed in is already the firstCol, then reverse its direction
     * If a new column is passed in, then make the existing col the 2nd sort order
     * @param {?} colName
     * @return {?}
     */
    function (colName) {
        if (this.sortParams.firstCol === colName) {
            if (this.sortParams.firstDir === SortDir.desc) {
                this.sortParams.firstDir = SortDir.asc;
                return;
            }
            else {
                this.sortParams.firstDir = SortDir.desc;
                return;
            }
        }
        else {
            this.sortParams.secondCol = this.sortParams.firstCol;
            this.sortParams.secondDir = this.sortParams.firstDir;
            this.sortParams.firstCol = colName;
            this.sortParams.firstDir = SortDir.desc;
        }
        this.log.debug('Sort params', this.sortParams);
        this.requestTableData();
    };
    /**
     * @param {?} column
     * @return {?}
     */
    TableBaseImpl.prototype.sortIcon = /**
     * @param {?} column
     * @return {?}
     */
    function (column) {
        if (this.sortParams.firstCol === column) {
            if (this.sortParams.firstDir === SortDir.asc) {
                return 'upArrow';
            }
            else {
                return 'downArrow';
            }
        }
        else {
            return '';
        }
    };
    /**
     * De-selects the row
     */
    /**
     * De-selects the row
     * @param {?} event
     * @return {?}
     */
    TableBaseImpl.prototype.deselectRow = /**
     * De-selects the row
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.log.debug('Details panel close event');
        this.selId = event;
    };
    return TableBaseImpl;
}());
/**
 * ONOS GUI -- Widget -- Table Base class
 * @abstract
 */
export { TableBaseImpl };
if (false) {
    /** @type {?} */
    TableBaseImpl.prototype.annots;
    /** @type {?} */
    TableBaseImpl.prototype.changedData;
    /** @type {?} */
    TableBaseImpl.prototype.payloadParams;
    /** @type {?} */
    TableBaseImpl.prototype.sortParams;
    /** @type {?} */
    TableBaseImpl.prototype.selectCallback;
    /** @type {?} */
    TableBaseImpl.prototype.parentSelCb;
    /** @type {?} */
    TableBaseImpl.prototype.responseCallback;
    /** @type {?} */
    TableBaseImpl.prototype.selId;
    /** @type {?} */
    TableBaseImpl.prototype.tableData;
    /** @type {?} */
    TableBaseImpl.prototype.tableDataFilter;
    /** @type {?} */
    TableBaseImpl.prototype.toggleRefresh;
    /** @type {?} */
    TableBaseImpl.prototype.autoRefresh;
    /** @type {?} */
    TableBaseImpl.prototype.autoRefreshTip;
    /** @type {?} */
    TableBaseImpl.prototype.root;
    /** @type {?} */
    TableBaseImpl.prototype.req;
    /** @type {?} */
    TableBaseImpl.prototype.resp;
    /** @type {?} */
    TableBaseImpl.prototype.refreshPromise;
    /** @type {?} */
    TableBaseImpl.prototype.handlers;
    /** @type {?} */
    TableBaseImpl.prototype.fs;
    /** @type {?} */
    TableBaseImpl.prototype.ls;
    /** @type {?} */
    TableBaseImpl.prototype.log;
    /** @type {?} */
    TableBaseImpl.prototype.wss;
    /** @type {?} */
    TableBaseImpl.prototype.tag;
    /** @type {?} */
    TableBaseImpl.prototype.idKey;
    /** @type {?} */
    TableBaseImpl.prototype.selCb;
}
