// gui2-fw-lib/esm5/lib/remote/urlfn.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
/*
 * Copyright 2018-present Open Networking Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Injectable, Inject } from '@angular/core';
import { LogService } from '../log.service';
import * as i0 from "@angular/core";
import * as i1 from "../log.service";
/** @type {?} */
var UICONTEXT = '/onos/ui/';
/** @type {?} */
var RSSUFFIX = UICONTEXT + 'rs/';
/** @type {?} */
var WSSUFFIX = UICONTEXT + 'websock/';
/**
 * ONOS GUI -- Remote -- General Purpose URL Functions
 */
var UrlFnService = /** @class */ (function () {
    function UrlFnService(log, w) {
        this.log = log;
        this.w = w;
        this.log.debug('UrlFnService constructed');
    }
    /**
     * @param {?} protocol
     * @return {?}
     */
    UrlFnService.prototype.matchSecure = /**
     * @param {?} protocol
     * @return {?}
     */
    function (protocol) {
        /** @type {?} */
        var p = this.w.location.protocol;
        /** @type {?} */
        var secure = (p === 'https' || p === 'wss');
        return secure ? protocol + 's' : protocol;
    };
    /* A little bit of funky here. It is possible that ONOS sits
     * behind a proxy and has an app prefix, e.g.
     *      http://host:port/my/app/onos/ui...
     * This bit of regex grabs everything after the host:port and
     * before the UICONTEXT (/onos/ui/) and uses that as an app
     * prefix by inserting it back into the WS URL.
     * If no prefix, then no insert.
     */
    /* A little bit of funky here. It is possible that ONOS sits
         * behind a proxy and has an app prefix, e.g.
         *      http://host:port/my/app/onos/ui...
         * This bit of regex grabs everything after the host:port and
         * before the UICONTEXT (/onos/ui/) and uses that as an app
         * prefix by inserting it back into the WS URL.
         * If no prefix, then no insert.
         */
    /**
     * @param {?} protocol
     * @param {?=} port
     * @param {?=} host
     * @return {?}
     */
    UrlFnService.prototype.urlBase = /* A little bit of funky here. It is possible that ONOS sits
         * behind a proxy and has an app prefix, e.g.
         *      http://host:port/my/app/onos/ui...
         * This bit of regex grabs everything after the host:port and
         * before the UICONTEXT (/onos/ui/) and uses that as an app
         * prefix by inserting it back into the WS URL.
         * If no prefix, then no insert.
         */
    /**
     * @param {?} protocol
     * @param {?=} port
     * @param {?=} host
     * @return {?}
     */
    function (protocol, port, host) {
        if (port === void 0) { port = ''; }
        if (host === void 0) { host = ''; }
        /** @type {?} */
        var match = this.w.location.href.match('.*//[^/]+/(.+)' + UICONTEXT);
        /** @type {?} */
        var appPrefix = match ? '/' + match[1] : '';
        return this.matchSecure(protocol) +
            '://' +
            (host === '' ? this.w.location.hostname : host) +
            ':' +
            (port === '' ? this.w.location.port : port) +
            appPrefix;
    };
    /**
     * @param {?} suffix
     * @return {?}
     */
    UrlFnService.prototype.httpPrefix = /**
     * @param {?} suffix
     * @return {?}
     */
    function (suffix) {
        return this.urlBase('http') + suffix;
    };
    /**
     * @param {?} suffix
     * @param {?} wsport
     * @param {?} host
     * @return {?}
     */
    UrlFnService.prototype.wsPrefix = /**
     * @param {?} suffix
     * @param {?} wsport
     * @param {?} host
     * @return {?}
     */
    function (suffix, wsport, host) {
        return this.urlBase('ws', wsport, host) + suffix;
    };
    /**
     * @param {?} path
     * @return {?}
     */
    UrlFnService.prototype.rsUrl = /**
     * @param {?} path
     * @return {?}
     */
    function (path) {
        return this.httpPrefix(RSSUFFIX) + path;
    };
    /**
     * @param {?} path
     * @param {?=} wsport
     * @param {?=} host
     * @return {?}
     */
    UrlFnService.prototype.wsUrl = /**
     * @param {?} path
     * @param {?=} wsport
     * @param {?=} host
     * @return {?}
     */
    function (path, wsport, host) {
        return this.wsPrefix(WSSUFFIX, wsport, host) + path;
    };
    UrlFnService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                },] },
    ];
    /** @nocollapse */
    UrlFnService.ctorParameters = function () { return [
        { type: LogService },
        { type: undefined, decorators: [{ type: Inject, args: ['Window',] }] }
    ]; };
    /** @nocollapse */ UrlFnService.ngInjectableDef = i0.defineInjectable({ factory: function UrlFnService_Factory() { return new UrlFnService(i0.inject(i1.LogService), i0.inject("Window")); }, token: UrlFnService, providedIn: "root" });
    return UrlFnService;
}());
export { UrlFnService };
if (false) {
    /** @type {?} */
    UrlFnService.prototype.log;
    /** @type {?} */
    UrlFnService.prototype.w;
}
