// gui2-fw-lib/esm5/lib/layer/veil/veil.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
/*
 * Copyright 2018-present Open Networking Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Component } from '@angular/core';
import { FnService } from '../../util/fn.service';
import { GlyphService } from '../../svg/glyph.service';
import { LogService } from '../../log.service';
import { SvgUtilService } from '../../svg/svgutil.service';
import { WebSocketService } from '../../remote/websocket.service';
/** @type {?} */
var BIRD = 'bird';
/**
 * ONOS GUI -- Layer -- Veil Component
 *
 * Provides a mechanism to display an overlaying div with information.
 * Used mainly for web socket connection interruption.
 *
 * It can be added to an component's template as follows:
 *     <onos-veil #veil></onos-veil>
 *     <p (click)="veil.show(['t1','t2','t3'])">Test Veil</p>
 */
var VeilComponent = /** @class */ (function () {
    function VeilComponent(fs, gs, log, sus, wss) {
        this.fs = fs;
        this.gs = gs;
        this.log = log;
        this.sus = sus;
        this.wss = wss;
        this.enabled = false;
        this.messages = [];
        /** @type {?} */
        var wSize = this.fs.windowSize();
        this.ww = wSize.width;
        this.wh = wSize.height;
        /** @type {?} */
        var shrink = this.wh * 0.3;
        this.birdDim = this.wh - shrink;
        /** @type {?} */
        var birdCenter = (this.ww - this.birdDim) / 2;
        this.trans = this.sus.translate([birdCenter, shrink / 2]);
        this.log.debug('VeilComponent with ' + BIRD + ' constructed');
    }
    /**
     * @return {?}
     */
    VeilComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    // msg should be an array of strings
    // msg should be an array of strings
    /**
     * @param {?} msgs
     * @return {?}
     */
    VeilComponent.prototype.show = 
    // msg should be an array of strings
    /**
     * @param {?} msgs
     * @return {?}
     */
    function (msgs) {
        this.messages = msgs;
        this.enabled = true;
        //        this.ks.enableKeys(false);
    };
    /**
     * @return {?}
     */
    VeilComponent.prototype.hide = /**
     * @return {?}
     */
    function () {
        this.veilStyle = 'display: none';
        //        this.ks.enableKeys(true);
    };
    VeilComponent.decorators = [
        { type: Component, args: [{
                    selector: 'onos-veil',
                    template: "<!--\n~ Copyright 2018-present Open Networking Foundation\n~\n~ Licensed under the Apache License, Version 2.0 (the \"License\");\n~ you may not use this file except in compliance with the License.\n~ You may obtain a copy of the License at\n~\n~     http://www.apache.org/licenses/LICENSE-2.0\n~\n~ Unless required by applicable law or agreed to in writing, software\n~ distributed under the License is distributed on an \"AS IS\" BASIS,\n~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n~ See the License for the specific language governing permissions and\n~ limitations under the License.\n-->\n\n<div id=\"veil\" *ngIf=\"enabled\">\n    <p *ngFor=\"let msg of messages\">{{ msg }}</p>\n    <svg [attr.width]=\"fs.windowSize().width\" [attr.height]=\"fs.windowSize().height\">\n        <use [attr.width]=\"birdDim\" [attr.height]=\"birdDim\" class=\"glyph\"\n             style=\"opacity: 0.2;\"\n             xlink:href = \"#bird\" [attr.transform]=\"trans\"/>\n\n    </svg>\n</div>\n",
                    styles: ["#veil{z-index:5000;display:block;position:absolute;top:0;left:0;padding:60px}#veil p{display:block;text-align:left;font-size:14pt;font-style:italic}", "#veil,.light{background-color:rgba(0,0,0,.75)}#veil,.dark{background-color:rgba(64,64,64,.75)}#veil p{color:#ddd}#veil svg .glyph{fill:#222}"]
                },] },
    ];
    /** @nocollapse */
    VeilComponent.ctorParameters = function () { return [
        { type: FnService },
        { type: GlyphService },
        { type: LogService },
        { type: SvgUtilService },
        { type: WebSocketService }
    ]; };
    return VeilComponent;
}());
export { VeilComponent };
if (false) {
    /** @type {?} */
    VeilComponent.prototype.ww;
    /** @type {?} */
    VeilComponent.prototype.wh;
    /** @type {?} */
    VeilComponent.prototype.birdSvg;
    /** @type {?} */
    VeilComponent.prototype.birdDim;
    /** @type {?} */
    VeilComponent.prototype.enabled;
    /** @type {?} */
    VeilComponent.prototype.trans;
    /** @type {?} */
    VeilComponent.prototype.messages;
    /** @type {?} */
    VeilComponent.prototype.veilStyle;
    /** @type {?} */
    VeilComponent.prototype.fs;
    /** @type {?} */
    VeilComponent.prototype.gs;
    /** @type {?} */
    VeilComponent.prototype.log;
    /** @type {?} */
    VeilComponent.prototype.sus;
    /** @type {?} */
    VeilComponent.prototype.wss;
}
