// gui2-fw-lib/esm5/lib/consolelogger.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
/*
 * Copyright 2018-present Open Networking Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Injectable } from '@angular/core';
import { environment } from '../environments/environment';
import * as i0 from "@angular/core";
/** @type {?} */
export var isDebugMode = !environment.production;
/** @type {?} */
var noop = function () { return undefined; };
var ɵ0 = noop;
/**
 * ONOS GUI -- LogService
 * Inspired by https://robferguson.org/blog/2017/09/09/a-simple-logging-service-for-angular-4/
 */
var ConsoleLoggerService = /** @class */ (function () {
    function ConsoleLoggerService() {
    }
    Object.defineProperty(ConsoleLoggerService.prototype, "debug", {
        get: /**
         * @return {?}
         */
        function () {
            if (isDebugMode) {
                return console.debug.bind(console);
            }
            else {
                return noop;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ConsoleLoggerService.prototype, "info", {
        get: /**
         * @return {?}
         */
        function () {
            if (isDebugMode) {
                return console.info.bind(console);
            }
            else {
                return noop;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ConsoleLoggerService.prototype, "warn", {
        get: /**
         * @return {?}
         */
        function () {
            return console.warn.bind(console);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ConsoleLoggerService.prototype, "error", {
        get: /**
         * @return {?}
         */
        function () {
            return console.error.bind(console);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} type
     * @param {?=} args
     * @return {?}
     */
    ConsoleLoggerService.prototype.invokeConsoleMethod = /**
     * @param {?} type
     * @param {?=} args
     * @return {?}
     */
    function (type, args) {
        /** @type {?} */
        var logFn = (console)[type] || console.log || noop;
        logFn.apply(console, [args]);
    };
    ConsoleLoggerService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                },] },
    ];
    /** @nocollapse */ ConsoleLoggerService.ngInjectableDef = i0.defineInjectable({ factory: function ConsoleLoggerService_Factory() { return new ConsoleLoggerService(); }, token: ConsoleLoggerService, providedIn: "root" });
    return ConsoleLoggerService;
}());
export { ConsoleLoggerService };
export { ɵ0 };
