// gui2-fw-lib/esm5/lib/mast/mast.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
/*
 * Copyright 2018-present Open Networking Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Injectable } from '@angular/core';
import { FnService } from '../util/fn.service';
import { LogService } from '../log.service';
import * as i0 from "@angular/core";
import * as i1 from "../util/fn.service";
import * as i2 from "../log.service";
/** @type {?} */
var PADMOBILE = 16;
/**
 * ONOS GUI -- Masthead Service
 */
var MastService = /** @class */ (function () {
    function MastService(fs, log) {
        this.fs = fs;
        this.log = log;
        this.mastHeight = 48;
        if (this.fs.isMobile()) {
            this.mastHeight += PADMOBILE;
        }
        this.log.debug('MastService constructed');
    }
    MastService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                },] },
    ];
    /** @nocollapse */
    MastService.ctorParameters = function () { return [
        { type: FnService },
        { type: LogService }
    ]; };
    /** @nocollapse */ MastService.ngInjectableDef = i0.defineInjectable({ factory: function MastService_Factory() { return new MastService(i0.inject(i1.FnService), i0.inject(i2.LogService)); }, token: MastService, providedIn: "root" });
    return MastService;
}());
export { MastService };
if (false) {
    /** @type {?} */
    MastService.prototype.mastHeight;
    /** @type {?} */
    MastService.prototype.fs;
    /** @type {?} */
    MastService.prototype.log;
}
