// gui2-fw-lib/esm5/lib/util/lion.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/*
 * Copyright 2018-present Open Networking Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
import { Injectable } from '@angular/core';
import { LogService } from '../log.service';
import { WebSocketService } from '../remote/websocket.service';
import * as i0 from "@angular/core";
import * as i1 from "../log.service";
import * as i2 from "../remote/websocket.service";
/**
 * A definition of Lion data
 * @record
 */
export function Lion() { }
if (false) {
    /** @type {?} */
    Lion.prototype.locale;
    /** @type {?} */
    Lion.prototype.lion;
}
/**
 * ONOS GUI -- Lion -- Localization Utilities
 */
var LionService = /** @class */ (function () {
    function LionService(log, wss) {
        var _this = this;
        this.log = log;
        this.wss = wss;
        this.ubercache = [];
        this.loadCbs = new Map([]); // A map of functions
        this.wss.bindHandlers(new Map([
            ['uberlion', function (data) { return _this.uberlion(data); }]
        ]));
        this.log.debug('LionService constructed');
    }
    /**
     * Handler for uberlion event from WSS
     */
    // A map of functions
    /**
     * Handler for uberlion event from WSS
     * @param {?} data
     * @return {?}
     */
    LionService.prototype.uberlion = 
    // A map of functions
    /**
     * Handler for uberlion event from WSS
     * @param {?} data
     * @return {?}
     */
    function (data) {
        var e_1, _a;
        this.ubercache = data.lion;
        this.log.info('LION service: Locale... [' + data.locale + ']');
        this.log.info('LION service: Bundles installed...');
        for (var p in this.ubercache) {
            if (this.ubercache[p]) {
                this.log.info('            :=> ', p);
            }
        }
        try {
            // If any component had registered a callback, call it now
            // that LION is loaded
            for (var _b = tslib_1.__values(this.loadCbs.keys()), _c = _b.next(); !_c.done; _c = _b.next()) {
                var cbname = _c.value;
                this.log.debug('Updating', cbname, 'with LION');
                this.loadCbs.get(cbname)();
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        this.log.debug('LION service: uber-lion bundle received:', data);
    };
    /**
     * Returns a lion bundle (function) for the given bundle ID (string)
     * returns a function that takes a string and returns a string
     */
    /**
     * Returns a lion bundle (function) for the given bundle ID (string)
     * returns a function that takes a string and returns a string
     * @param {?} bundleId
     * @return {?}
     */
    LionService.prototype.bundle = /**
     * Returns a lion bundle (function) for the given bundle ID (string)
     * returns a function that takes a string and returns a string
     * @param {?} bundleId
     * @return {?}
     */
    function (bundleId) {
        /** @type {?} */
        var bundleObj = this.ubercache[bundleId];
        if (!bundleObj) {
            this.log.warn('No lion bundle registered:', bundleId);
            bundleObj = {};
        }
        return function (key) {
            return bundleObj[key] || '%' + key + '%';
        };
    };
    LionService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                },] },
    ];
    /** @nocollapse */
    LionService.ctorParameters = function () { return [
        { type: LogService },
        { type: WebSocketService }
    ]; };
    /** @nocollapse */ LionService.ngInjectableDef = i0.defineInjectable({ factory: function LionService_Factory() { return new LionService(i0.inject(i1.LogService), i0.inject(i2.WebSocketService)); }, token: LionService, providedIn: "root" });
    return LionService;
}());
export { LionService };
if (false) {
    /** @type {?} */
    LionService.prototype.ubercache;
    /** @type {?} */
    LionService.prototype.loadCbs;
    /** @type {?} */
    LionService.prototype.log;
    /** @type {?} */
    LionService.prototype.wss;
}
