// gui2-fw-lib/esm5/lib/util/theme.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
/*
 * Copyright 2018-present Open Networking Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Injectable } from '@angular/core';
import { LogService } from '../log.service';
import * as i0 from "@angular/core";
import * as i1 from "../log.service";
/**
 * ONOS GUI -- Util -- Theme Service
 */
var ThemeService = /** @class */ (function () {
    function ThemeService(log) {
        this.log = log;
        this.themes = ['light', 'dark'];
        this.thidx = 0;
        this.log.debug('ThemeService constructed');
    }
    /**
     * @return {?}
     */
    ThemeService.prototype.getTheme = /**
     * @return {?}
     */
    function () {
        return this.themes[this.thidx];
    };
    /**
     * @return {?}
     */
    ThemeService.prototype.themeStr = /**
     * @return {?}
     */
    function () {
        return this.themes.join(' ');
    };
    ThemeService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                },] },
    ];
    /** @nocollapse */
    ThemeService.ctorParameters = function () { return [
        { type: LogService }
    ]; };
    /** @nocollapse */ ThemeService.ngInjectableDef = i0.defineInjectable({ factory: function ThemeService_Factory() { return new ThemeService(i0.inject(i1.LogService)); }, token: ThemeService, providedIn: "root" });
    return ThemeService;
}());
export { ThemeService };
if (false) {
    /** @type {?} */
    ThemeService.prototype.themes;
    /** @type {?} */
    ThemeService.prototype.thidx;
    /** @type {?} */
    ThemeService.prototype.log;
}
