// gui2-fw-lib/esm5/lib/util/trie.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
/*
 * Copyright 2018-present Open Networking Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * @record
 */
export function TrieC() { }
if (false) {
    /** @type {?} */
    TrieC.prototype.p;
    /** @type {?} */
    TrieC.prototype.s;
}
/**
 * @record
 */
export function TrieT() { }
if (false) {
    /** @type {?} */
    TrieT.prototype.k;
    /** @type {?} */
    TrieT.prototype.p;
    /** @type {?} */
    TrieT.prototype.q;
}
/** @enum {string} */
var TrieRemoved = {
    REMOVED: 'removed',
    ABSENT: 'absent',
};
export { TrieRemoved };
/** @enum {string} */
var TrieInsert = {
    ADDED: 'added',
    UPDATED: 'updated',
};
export { TrieInsert };
/** @enum {string} */
var TrieOp = {
    PLUS: '+',
    MINUS: '-',
};
export { TrieOp };
var Trie = /** @class */ (function () {
    function Trie(op, trie, word, data) {
        this.p = trie;
        this.w = word.toUpperCase();
        this.s = this.w.split('');
        this.c = { p: this.p, s: this.s },
            this.t = [];
        this.x = 0;
        this.f1 = op === TrieOp.PLUS ? this.add : this.probe;
        this.f2 = op === TrieOp.PLUS ? this.insert : this.remove;
        this.data = data;
        while (this.c.s.length) {
            this.c = this.f1(this.c);
        }
    }
    /**
     * @param {?} cAdded
     * @return {?}
     */
    Trie.prototype.add = /**
     * @param {?} cAdded
     * @return {?}
     */
    function (cAdded) {
        /** @type {?} */
        var q = cAdded.s.shift();
        /** @type {?} */
        var np = cAdded.p[q];
        if (!np) {
            cAdded.p[q] = {};
            np = cAdded.p[q];
            this.x = 1;
        }
        return { p: np, s: cAdded.s };
    };
    /**
     * @param {?} cProbed
     * @return {?}
     */
    Trie.prototype.probe = /**
     * @param {?} cProbed
     * @return {?}
     */
    function (cProbed) {
        /** @type {?} */
        var q = cProbed.s.shift();
        /** @type {?} */
        var k = Object.keys(cProbed.p).length;
        /** @type {?} */
        var np = cProbed.p[q];
        this.t.push({ q: q, k: k, p: cProbed.p });
        if (!np) {
            this.t = [];
            return { p: [], s: [] };
        }
        return { p: np, s: cProbed.s };
    };
    /**
     * @return {?}
     */
    Trie.prototype.insert = /**
     * @return {?}
     */
    function () {
        this.c.p._data = this.data;
        return this.x ? TrieInsert.ADDED : TrieInsert.UPDATED;
    };
    /**
     * @return {?}
     */
    Trie.prototype.remove = /**
     * @return {?}
     */
    function () {
        if (this.t.length) {
            this.t = this.t.reverse();
            while (this.t.length) {
                /** @type {?} */
                var d = this.t.shift();
                delete d.p[d.q];
                if (d.k > 1) {
                    this.t = [];
                }
            }
            return TrieRemoved.REMOVED;
        }
        return TrieRemoved.ABSENT;
    };
    return Trie;
}());
export { Trie };
if (false) {
    /** @type {?} */
    Trie.prototype.p;
    /** @type {?} */
    Trie.prototype.w;
    /** @type {?} */
    Trie.prototype.s;
    /** @type {?} */
    Trie.prototype.c;
    /** @type {?} */
    Trie.prototype.t;
    /** @type {?} */
    Trie.prototype.x;
    /** @type {?} */
    Trie.prototype.f1;
    /** @type {?} */
    Trie.prototype.f2;
    /** @type {?} */
    Trie.prototype.data;
}
