// gui2-fw-lib/esm5/lib/layer/quickhelp/quickhelp.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
/*
 * Copyright 2019-present Open Networking Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Component } from '@angular/core';
import { animate, state, style, transition, trigger } from '@angular/animations';
import { LogService } from '../../log.service';
import { FnService } from '../../util/fn.service';
import { KeysService } from '../../util/keys.service';
import { LionService } from '../../util/lion.service';
var QuickhelpComponent = /** @class */ (function () {
    function QuickhelpComponent(log, fs, ks, lion) {
        var _this = this;
        this.log = log;
        this.fs = fs;
        this.ks = ks;
        this.lion = lion;
        if (this.lion.ubercache.length === 0) {
            this.lionFn = this.dummyLion;
            this.lion.loadCbs.set('quickhelp', function () { return _this.doLion(); });
        }
        else {
            this.doLion();
        }
        this.log.debug('Quickhelp component constructed');
    }
    /**
     * Read the LION bundle for Toolbar and set up the lionFn
     */
    /**
     * Read the LION bundle for Toolbar and set up the lionFn
     * @return {?}
     */
    QuickhelpComponent.prototype.doLion = /**
     * Read the LION bundle for Toolbar and set up the lionFn
     * @return {?}
     */
    function () {
        this.lionFn = this.lion.bundle('core.fw.QuickHelp');
    };
    /**
    * A dummy implementation of the lionFn until the response is received and the LION
    * bundle is received from the WebSocket
    */
    /**
     * A dummy implementation of the lionFn until the response is received and the LION
     * bundle is received from the WebSocket
     * @param {?} key
     * @return {?}
     */
    QuickhelpComponent.prototype.dummyLion = /**
     * A dummy implementation of the lionFn until the response is received and the LION
     * bundle is received from the WebSocket
     * @param {?} key
     * @return {?}
     */
    function (key) {
        return '%' + key + '%';
    };
    QuickhelpComponent.decorators = [
        { type: Component, args: [{
                    selector: 'onos-quickhelp',
                    template: "<!--\n~ Copyright 2019-present Open Networking Foundation\n~\n~ Licensed under the Apache License, Version 2.0 (the \"License\");\n~ you may not use this file except in compliance with the License.\n~ You may obtain a copy of the License at\n~\n~     http://www.apache.org/licenses/LICENSE-2.0\n~\n~ Unless required by applicable law or agreed to in writing, software\n~ distributed under the License is distributed on an \"AS IS\" BASIS,\n~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n~ See the License for the specific language governing permissions and\n~ limitations under the License.\n-->\n<div id=\"quickhelp\" [@quickHelpState]=\"ks.quickHelpShown\">\n    <div class=\"help\" *ngIf=\"ks.quickHelpShown\">\n        <p class=\"title\">{{lionFn(\"qh_title\")}}</p>\n        <!-- TODO: drive this through the keys service keyHandler -->\n        <table class=\"qhrow\">\n            <tr>\n                <td class=\"key\">&#92;</td>\n                <td class=\"desc\">Afficher/cacher l'aide rapide</td>\n            </tr>\n            <tr>\n                <td class=\"key\">/</td>\n                <td class=\"desc\">Afficher/cacher l'aide rapide</td>\n            </tr>\n            <tr>\n                <td class=\"key\">Esc</td>\n                <td class=\"desc\">ignorer la bo\u00EEte de dialogue ou annuler les\n                    s\u00E9lections\n                </td>\n            </tr>\n            <tr>\n                <td class=\"key\">T</td>\n                <td class=\"desc\">basculer de th\u00E8me</td>\n            </tr>\n        </table>\n        <hr class=\"qhrowsep\">\n        <table class=\"qhrow\">\n            <tr>\n                <td class=\"key\">I</td>\n                <td class=\"desc\">Basculer au panneau de instance\n                    ONOS\n                </td>\n                <td class=\"key\">O</td>\n                <td class=\"desc\">Baculer au panneau r\u00E9capitulatif\n                    ONOS\n                </td>\n                <td class=\"key\">D</td>\n                <td class=\"desc\">Basculer le panneau de d\u00E9tails</td>\n            </tr>\n            <tr>\n                <td class=\"key\">H</td>\n                <td class=\"desc\">Basculer la visibilit\u00E9 de l'h\u00F4te\n                </td>\n                <td class=\"key\">M</td>\n                <td class=\"desc\">Basculer la visibilit\u00E9 hors-ligne\n                </td>\n                <td class=\"key\">P</td>\n                <td class=\"desc\">Basculer le port mis en \u00E9vidence\n                </td>\n            </tr>\n            <tr>\n                <td class=\"key\">Dash</td>\n                <td class=\"desc\">Montrer les mauvais liens</td>\n                <td class=\"key\">B</td>\n                <td class=\"desc\">Basculer la carte g\u00E9o de fond</td>\n                <td class=\"key\">G</td>\n                <td class=\"desc\">S\u00E9lectionner la carte g\u00E9o de fond\n                </td>\n            </tr>\n            <tr>\n                <td class=\"key\">S</td>\n                <td class=\"desc\">Toggle sprite layer</td>\n                <td class=\"key\" y=\"0\">X</td>\n                <td class=\"desc\" y=\"0\" x=\"33.53125\">r\u00E9initialiser la\n                    carte des noeuds\n                </td>\n                <td class=\"key\">Z</td>\n                <td class=\"desc\">Basculer la vue oblique\n                    (exp\u00E9rimental)\n                </td>\n            </tr>\n            <tr>\n                <td class=\"key\">N</td>\n                <td class=\"desc\">cycle couche de noeuds</td>\n                <td class=\"key\">L</td>\n                <td class=\"desc\">Parcourir \u00E9tiquettes appareils</td>\n                <td class=\"key\">Shift-l</td>\n                <td class=\"desc\">Parcourir \u00E9tiquettes h\u00F4te</td>\n            </tr>\n            <tr>\n                <td class=\"key\">U</td>\n                <td class=\"desc\">survoler avec la souris</td>\n                <td class=\"key\">R</td>\n                <td class=\"desc\">r\u00E9initialiser le zoom et le\n                    panoramique\n                </td>\n                <td class=\"key\">E</td>\n                <td class=\"desc\">Egaliser les r\u00F4les de ma\u00EEtrises\n                </td>\n            </tr>\n            <tr>\n                <td class=\"key\">Dot</td>\n                <td class=\"desc\">Basculer a la barre d'outils</td>\n                <td class=\"key\">0</td>\n                <td class=\"desc\">Annuler la surveillance du trafic\n                </td>\n                <td class=\"key\">A</td>\n                <td class=\"desc\">Surveiller tout le trafic</td>\n            </tr>\n            <tr>\n                <td class=\"key\">F</td>\n                <td class=\"desc\">Afficher les flux de liaison de p\u00E9riph\u00E9rique\n                </td>\n                <td class=\"key\">V</td>\n                <td class=\"desc\">Show all related intents</td>\n                <td class=\"key\">L-arrow</td>\n                <td class=\"desc\">Show previous related intent</td>\n            </tr>\n            <tr>\n                <td class=\"key\">R-arrow</td>\n                <td class=\"desc\">Show next related intent</td>\n                <td class=\"key\">W</td>\n                <td class=\"desc\">Monitor traffic of selected intent</td>\n            </tr>\n        </table>\n        <hr>\n        <div class=\"qhrow\">\n            <table class=\"qh-r4-c0\">\n                <tr>\n                    <td class=\"key\">cliquer</td>\n                    <td class=\"desc\">S\u00E9lectionner l'article et montrer les\n                        d\u00E9tails\n                    </td>\n                </tr>\n                <tr>\n                    <td class=\"key\">Cliquer sur MAJ</td>\n                    <td class=\"desc\">Basculer l'\u00E9tat de s\u00E9lection</td>\n                </tr>\n                <tr>\n                    <td class=\"key\">Glisser</td>\n                    <td class=\"desc\">Repositionner (et \u00E9pingler) le p\u00E9riph\u00E9rique\n                        / l'h\u00F4te\n                    </td>\n                </tr>\n                <tr>\n                    <td class=\"key\">cmd d\u00E9filer</td>\n                    <td class=\"desc\">Zoom avant/arri\u00E8re</td>\n                </tr>\n                <tr>\n                    <td class=\"key\" y=\"48\">cmd-glisser</td>\n                    <td class=\"desc\" y=\"48\" x=\"74.84375\">Pan\n                </tr>\n            </table>\n        </div>\n    </div>\n</div>\n",
                    styles: ["#quickhelp{top:100px;z-index:5000;position:relative}#quickhelp div.help{background:#000;opacity:.8}#quickhelp div.help table{vertical-align:top}#quickhelp div.help p{text-align:center;color:#fff;font-weight:700}#quickhelp td.key{color:#add;padding-left:8px;padding-right:4px}#quickhelp td.desc{color:#fff}"],
                    animations: [
                        trigger('quickHelpState', [
                            state('true', style({
                                opacity: '1.0',
                            })),
                            state('false', style({
                                opacity: '0.0',
                            })),
                            transition('0 => 1', animate('500ms ease-in')),
                            transition('1 => 0', animate('500ms ease-out'))
                        ])
                    ]
                },] },
    ];
    /** @nocollapse */
    QuickhelpComponent.ctorParameters = function () { return [
        { type: LogService },
        { type: FnService },
        { type: KeysService },
        { type: LionService }
    ]; };
    return QuickhelpComponent;
}());
export { QuickhelpComponent };
if (false) {
    /** @type {?} */
    QuickhelpComponent.prototype.lionFn;
    /** @type {?} */
    QuickhelpComponent.prototype.log;
    /** @type {?} */
    QuickhelpComponent.prototype.fs;
    /** @type {?} */
    QuickhelpComponent.prototype.ks;
    /** @type {?} */
    QuickhelpComponent.prototype.lion;
}
