// gui2-fw-lib/esm2015/lib/nav/nav.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
/*
 * Copyright 2018-present Open Networking Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { FnService } from '../util/fn.service';
import { LogService } from '../log.service';
import * as i0 from "@angular/core";
import * as i1 from "../util/fn.service";
import * as i2 from "../log.service";
import * as i3 from "@angular/common/http";
/**
 * @record
 */
export function UiView() { }
if (false) {
    /** @type {?} */
    UiView.prototype.id;
    /** @type {?} */
    UiView.prototype.icon;
    /** @type {?} */
    UiView.prototype.cat;
    /** @type {?} */
    UiView.prototype.label;
}
/**
 * ONOS GUI -- Navigation Service
 */
export class NavService {
    /**
     * @param {?} _fn_
     * @param {?} log
     * @param {?} httpClient
     */
    constructor(_fn_, log, httpClient) {
        this._fn_ = _fn_;
        this.log = log;
        this.httpClient = httpClient;
        this.showNav = false;
        this.uiPlatformViews = new Array();
        this.uiNetworkViews = new Array();
        this.uiOtherViews = new Array();
        this.uiHiddenViews = new Array();
        this.log.debug('NavService constructed');
    }
    /**
     * @return {?}
     */
    hideNav() {
        this.showNav = false;
        this.log.debug('Hiding Nav menu');
    }
    /**
     * @return {?}
     */
    toggleNav() {
        this.showNav = !this.showNav;
        if (this.showNav) {
            this.log.debug('Showing Nav menu');
        }
        else {
            this.log.debug('Hiding Nav menu');
        }
    }
    /**
     * @return {?}
     */
    getUiViews() {
        this.uiPlatformViews = new Array();
        this.uiNetworkViews = new Array();
        this.uiOtherViews = new Array();
        this.uiHiddenViews = new Array();
        this.httpClient.get('rs/nav/uiextensions').subscribe((v) => {
            v.forEach((uiView) => {
                if (uiView.cat === 'PLATFORM') {
                    this.uiPlatformViews.push(uiView);
                }
                else if (uiView.cat === 'NETWORK') {
                    if (uiView.id !== 'topo') {
                        this.uiNetworkViews.push(uiView);
                    }
                    else {
                        this.uiNetworkViews.push((/** @type {?} */ ({
                            id: 'topo2',
                            icon: 'nav_topo',
                            cat: 'NETWORK',
                            label: uiView.label
                        })));
                    }
                }
                else if (uiView.cat === 'HIDDEN') {
                    this.uiHiddenViews.push(uiView);
                }
                else {
                    this.uiOtherViews.push(uiView);
                }
            });
        });
    }
}
NavService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] },
];
/** @nocollapse */
NavService.ctorParameters = () => [
    { type: FnService },
    { type: LogService },
    { type: HttpClient }
];
/** @nocollapse */ NavService.ngInjectableDef = i0.defineInjectable({ factory: function NavService_Factory() { return new NavService(i0.inject(i1.FnService), i0.inject(i2.LogService), i0.inject(i3.HttpClient)); }, token: NavService, providedIn: "root" });
if (false) {
    /** @type {?} */
    NavService.prototype.showNav;
    /** @type {?} */
    NavService.prototype.uiPlatformViews;
    /** @type {?} */
    NavService.prototype.uiNetworkViews;
    /** @type {?} */
    NavService.prototype.uiOtherViews;
    /** @type {?} */
    NavService.prototype.uiHiddenViews;
    /** @type {?} */
    NavService.prototype._fn_;
    /** @type {?} */
    NavService.prototype.log;
    /** @type {?} */
    NavService.prototype.httpClient;
}
