// gui2-fw-lib/esm2015/lib/widget/detailspanel.base.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { PanelBaseImpl } from './panel.base';
import { Output, EventEmitter, Input } from '@angular/core';
/**
 * A generic model of the data returned from the *DetailsResponse
 * @record
 */
export function DetailsResponse() { }
if (false) {
    /** @type {?} */
    DetailsResponse.prototype.details;
}
/**
 * Extends the PanelBase abstract class specifically for showing details
 *
 * This makes another call through WSS to the server for specific
 * details to fill the panel with
 *
 * This replaces the detailspanel service in the old gui
 * @abstract
 */
export class DetailsPanelBaseImpl extends PanelBaseImpl {
    /**
     * @param {?} fs
     * @param {?} ls
     * @param {?} log
     * @param {?} wss
     * @param {?} tag
     */
    constructor(fs, ls, log, wss, tag) {
        super(fs, ls, log);
        this.fs = fs;
        this.ls = ls;
        this.log = log;
        this.wss = wss;
        this.tag = tag;
        this.closeEvent = new EventEmitter();
        this.handlers = [];
        this.detailsData = {};
        this.closed = false;
        this.root = tag + 's';
        this.req = tag + 'DetailsRequest';
        this.resp = tag + 'DetailsResponse';
    }
    /**
     * When the details panel is created set up a listener on
     * Web Socket for details responses
     * @return {?}
     */
    init() {
        this.wss.bindHandlers(new Map([
            [this.resp, (data) => this.detailsPanelResponseCb(data)]
        ]));
        this.handlers.push(this.resp);
    }
    /**
     * When the details panel is destroyed this should be called to
     * de-register from the WebSocket
     * @return {?}
     */
    destroy() {
        this.wss.unbindHandlers(this.handlers);
    }
    /**
     * A callback that executes when the details data that was requested
     * on WebSocketService arrives.
     * @param {?} data
     * @return {?}
     */
    detailsPanelResponseCb(data) {
        this.detailsData = data['details'];
    }
    /**
     * Details Panel Data Request - should be called whenever row id changes
     * @param {?} query
     * @return {?}
     */
    requestDetailsPanelData(query) {
        this.closed = false;
        // Do not send if the Web Socket hasn't opened
        if (this.wss.isConnected()) {
            if (this.fs.debugOn('panel')) {
                this.log.debug('Details panel data REQUEST:', this.req, query);
            }
            this.wss.sendEvent(this.req, query);
        }
    }
    /**
     * this should be called when the details panel close button is clicked
     * @return {?}
     */
    close() {
        this.closed = true;
        this.id = null;
        this.closeEvent.emit(this.id);
    }
}
DetailsPanelBaseImpl.propDecorators = {
    id: [{ type: Input }],
    closeEvent: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    DetailsPanelBaseImpl.prototype.id;
    /** @type {?} */
    DetailsPanelBaseImpl.prototype.closeEvent;
    /** @type {?} */
    DetailsPanelBaseImpl.prototype.root;
    /** @type {?} */
    DetailsPanelBaseImpl.prototype.req;
    /** @type {?} */
    DetailsPanelBaseImpl.prototype.resp;
    /** @type {?} */
    DetailsPanelBaseImpl.prototype.handlers;
    /** @type {?} */
    DetailsPanelBaseImpl.prototype.detailsData;
    /** @type {?} */
    DetailsPanelBaseImpl.prototype.closed;
    /** @type {?} */
    DetailsPanelBaseImpl.prototype.fs;
    /** @type {?} */
    DetailsPanelBaseImpl.prototype.ls;
    /** @type {?} */
    DetailsPanelBaseImpl.prototype.log;
    /** @type {?} */
    DetailsPanelBaseImpl.prototype.wss;
    /** @type {?} */
    DetailsPanelBaseImpl.prototype.tag;
}
