// gui2-fw-lib/esm2015/lib/widget/panel.base.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
/**
 * Base model of panel view - implemented by Panel components
 * @record
 */
export function PanelBase() { }
if (false) {
    /**
     * @param {?} cb
     * @return {?}
     */
    PanelBase.prototype.showPanel = function (cb) { };
    /**
     * @param {?} cb
     * @return {?}
     */
    PanelBase.prototype.hidePanel = function (cb) { };
    /**
     * @param {?} cb
     * @return {?}
     */
    PanelBase.prototype.togglePanel = function (cb) { };
    /**
     * @return {?}
     */
    PanelBase.prototype.panelIsVisible = function () { };
}
/**
 * ONOS GUI -- Widget -- Panel Base class
 *
 * Replacing the panel service in the old implementation
 * @abstract
 */
export class PanelBaseImpl {
    /**
     * @param {?} fs
     * @param {?} ls
     * @param {?} log
     */
    constructor(fs, ls, log) {
        this.fs = fs;
        this.ls = ls;
        this.log = log;
        //        this.log.debug('Panel base class constructed');
    }
    /**
     * @param {?} cb
     * @return {?}
     */
    showPanel(cb) {
        this.on = true;
    }
    /**
     * @param {?} cb
     * @return {?}
     */
    hidePanel(cb) {
        this.on = false;
    }
    /**
     * @param {?} cb
     * @return {?}
     */
    togglePanel(cb) {
        if (this.on) {
            this.hidePanel(cb);
        }
        else {
            this.showPanel(cb);
        }
        return this.on;
    }
    /**
     * @return {?}
     */
    panelIsVisible() {
        return this.on;
    }
    /**
     * A dummy implementation of the lionFn until the response is received and the LION
     * bundle is received from the WebSocket
     * @param {?} key
     * @return {?}
     */
    dummyLion(key) {
        return '%' + key + '%';
    }
}
if (false) {
    /** @type {?} */
    PanelBaseImpl.prototype.on;
    /** @type {?} */
    PanelBaseImpl.prototype.fs;
    /** @type {?} */
    PanelBaseImpl.prototype.ls;
    /** @type {?} */
    PanelBaseImpl.prototype.log;
}
