// gui2-fw-lib/esm2015/lib/widget/tableresize.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
/*
 * Copyright 2018-present Open Networking Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Directive, Inject } from '@angular/core';
import { FnService } from '../util/fn.service';
import { LogService } from '../log.service';
import { MastService } from '../mast/mast.service';
import { HostListener } from '@angular/core';
import * as d3 from 'd3';
/**
 * ONOS GUI -- Widget -- Table Resize Directive
 */
export class TableResizeDirective {
    /**
     * @param {?} fs
     * @param {?} log
     * @param {?} mast
     * @param {?} w
     */
    constructor(fs, log, mast, w) {
        this.fs = fs;
        this.log = log;
        this.mast = mast;
        this.w = w;
        this.pdg = 22;
        log.info('TableResizeDirective constructed');
    }
    /**
     * @return {?}
     */
    ngAfterContentChecked() {
        this.tables = {
            thead: d3.select('div.table-header').select('table'),
            tbody: d3.select('div.table-body').select('table')
        };
        this.windowSize(this.tables);
    }
    /**
     * @param {?} tables
     * @return {?}
     */
    windowSize(tables) {
        /** @type {?} */
        const wsz = this.fs.windowSize(0, 30);
        this.adjustTable(tables, wsz.width, wsz.height);
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onResize(event) {
        this.windowSize(this.tables);
        return {
            h: this.w.innerHeight,
            w: this.w.innerWidth
        };
    }
    /**
     * @param {?} tables
     * @param {?} width
     * @param {?} height
     * @return {?}
     */
    adjustTable(tables, width, height) {
        this._width(tables.thead, width + 'px');
        this._width(tables.tbody, width + 'px');
        this.setHeight(tables.thead, d3.select('div.table-body'), height);
    }
    /**
     * @param {?} elem
     * @param {?} width
     * @return {?}
     */
    _width(elem, width) {
        elem.style('width', width);
    }
    /**
     * @param {?} thead
     * @param {?} body
     * @param {?} height
     * @return {?}
     */
    setHeight(thead, body, height) {
        /** @type {?} */
        const h = height - (this.mast.mastHeight +
            this.fs.noPxStyle(d3.select('.tabular-header'), 'height') +
            this.fs.noPxStyle(thead, 'height') + this.pdg);
        body.style('height', h + 'px');
    }
}
TableResizeDirective.decorators = [
    { type: Directive, args: [{
                selector: '[onosTableResize]',
            },] },
];
/** @nocollapse */
TableResizeDirective.ctorParameters = () => [
    { type: FnService },
    { type: LogService },
    { type: MastService },
    { type: undefined, decorators: [{ type: Inject, args: ['Window',] }] }
];
TableResizeDirective.propDecorators = {
    onResize: [{ type: HostListener, args: ['window:resize', ['$event.target'],] }]
};
if (false) {
    /** @type {?} */
    TableResizeDirective.prototype.pdg;
    /** @type {?} */
    TableResizeDirective.prototype.tables;
    /** @type {?} */
    TableResizeDirective.prototype.fs;
    /** @type {?} */
    TableResizeDirective.prototype.log;
    /** @type {?} */
    TableResizeDirective.prototype.mast;
    /** @type {?} */
    TableResizeDirective.prototype.w;
}
