// gui2-fw-lib/esm2015/lib/consolelogger.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
/*
 * Copyright 2018-present Open Networking Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Injectable } from '@angular/core';
import { environment } from '../environments/environment';
import * as i0 from "@angular/core";
/** @type {?} */
export let isDebugMode = !environment.production;
/** @type {?} */
const noop = () => undefined;
const ɵ0 = noop;
/**
 * ONOS GUI -- LogService
 * Inspired by https://robferguson.org/blog/2017/09/09/a-simple-logging-service-for-angular-4/
 */
export class ConsoleLoggerService {
    /**
     * @return {?}
     */
    get debug() {
        if (isDebugMode) {
            return console.debug.bind(console);
        }
        else {
            return noop;
        }
    }
    /**
     * @return {?}
     */
    get info() {
        if (isDebugMode) {
            return console.info.bind(console);
        }
        else {
            return noop;
        }
    }
    /**
     * @return {?}
     */
    get warn() {
        return console.warn.bind(console);
    }
    /**
     * @return {?}
     */
    get error() {
        return console.error.bind(console);
    }
    /**
     * @param {?} type
     * @param {?=} args
     * @return {?}
     */
    invokeConsoleMethod(type, args) {
        /** @type {?} */
        const logFn = (console)[type] || console.log || noop;
        logFn.apply(console, [args]);
    }
}
ConsoleLoggerService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] },
];
/** @nocollapse */ ConsoleLoggerService.ngInjectableDef = i0.defineInjectable({ factory: function ConsoleLoggerService_Factory() { return new ConsoleLoggerService(); }, token: ConsoleLoggerService, providedIn: "root" });
export { ɵ0 };
