// gui2-fw-lib/esm2015/lib/detectbrowser.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
/*
 * Copyright 2018-present Open Networking Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Inject } from '@angular/core';
import { Directive } from '@angular/core';
import { FnService } from './util/fn.service';
import { LogService } from './log.service';
import { OnosService } from './onos.service';
/**
 * ONOS GUI -- Detect Browser Directive
 */
export class DetectBrowserDirective {
    /**
     * @param {?} fs
     * @param {?} log
     * @param {?} onos
     * @param {?} w
     */
    constructor(fs, log, onos, w) {
        this.fs = fs;
        this.log = log;
        this.onos = onos;
        this.w = w;
        /** @type {?} */
        const body = document.getElementsByTagName('body')[0];
        //        let body = d3.select('body');
        /** @type {?} */
        let browser = '';
        if (fs.isChrome()) {
            browser = 'chrome';
        }
        else if (fs.isChromeHeadless()) {
            browser = 'chromeheadless';
        }
        else if (fs.isSafari()) {
            browser = 'safari';
        }
        else if (fs.isFirefox()) {
            browser = 'firefox';
        }
        else {
            this.log.warn('Unknown browser. ', 'Vendor:', this.w.navigator.vendor, 'Agent:', this.w.navigator.userAgent);
            return;
        }
        body.classList.add(browser);
        //        body.classed(browser, true);
        this.onos.browser = browser;
        if (fs.isMobile()) {
            body.classList.add('mobile');
            this.onos.mobile = true;
        }
        //        this.log.debug('Detected browser is', fs.cap(browser));
    }
}
DetectBrowserDirective.decorators = [
    { type: Directive, args: [{
                selector: '[onosDetectBrowser]'
            },] },
];
/** @nocollapse */
DetectBrowserDirective.ctorParameters = () => [
    { type: FnService },
    { type: LogService },
    { type: OnosService },
    { type: undefined, decorators: [{ type: Inject, args: ['Window',] }] }
];
if (false) {
    /** @type {?} */
    DetectBrowserDirective.prototype.fs;
    /** @type {?} */
    DetectBrowserDirective.prototype.log;
    /** @type {?} */
    DetectBrowserDirective.prototype.onos;
    /** @type {?} */
    DetectBrowserDirective.prototype.w;
}
