// gui2-fw-lib/esm5/lib/layer/panel.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
/*
 *  Copyright 2018-present Open Networking Foundation
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
import { Injectable } from '@angular/core';
import { FnService } from '../util/fn.service';
import { LogService } from '../log.service';
import { ThemeService } from '../util/theme.service';
import { WebSocketService } from '../remote/websocket.service';
import * as d3 from 'd3';
import * as i0 from "@angular/core";
import * as i1 from "../util/fn.service";
import * as i2 from "../log.service";
import * as i3 from "../util/theme.service";
import * as i4 from "../remote/websocket.service";
/** @type {?} */
var fs;
/** @type {?} */
var defaultSettings = {
    edge: 'right',
    width: 200,
    margin: 20,
    hideMargin: 20,
    xtnTime: 750,
    fade: true,
};
/** @type {?} */
var panels;
/** @type {?} */
var panelLayer;
/**
 * @return {?}
 */
function init() {
    panelLayer = d3.select('div#floatpanels');
    panelLayer.text('');
    panels = {};
}
// helpers for panel
/**
 * @return {?}
 */
function noop() {
}
/**
 * @param {?} p
 * @return {?}
 */
function margin(p) {
    return p.settings.margin;
}
/**
 * @param {?} p
 * @return {?}
 */
function hideMargin(p) {
    return p.settings.hideMargin;
}
/**
 * @param {?} p
 * @param {?} what
 * @return {?}
 */
function noPx(p, what) {
    return Number(p.el.style(what).replace(/px$/, ''));
}
/**
 * @param {?} p
 * @return {?}
 */
function widthVal(p) {
    return noPx(p, 'width');
}
/**
 * @param {?} p
 * @return {?}
 */
function heightVal(p) {
    return noPx(p, 'height');
}
/**
 * @param {?} p
 * @return {?}
 */
function pxShow(p) {
    return margin(p) + 'px';
}
/**
 * @param {?} p
 * @return {?}
 */
function pxHide(p) {
    return (-hideMargin(p) - widthVal(p) - (noPx(p, 'padding') * 2)) + 'px';
}
/**
 * @param {?} id
 * @param {?} settings
 * @return {?}
 */
function makePanel(id, settings) {
    /** @type {?} */
    var p = {
        id: id,
        settings: settings,
        on: false,
        el: null,
    };
    /** @type {?} */
    var api = {
        show: showPanel,
        hide: hidePanel,
        toggle: togglePanel,
        empty: emptyPanel,
        append: appendPanel,
        width: panelWidth,
        height: panelHeight,
        bbox: panelBBox,
        isVisible: panelIsVisible,
        classed: classed,
        el: panelEl,
    };
    p.el = panelLayer.append('div')
        .attr('id', id)
        .attr('class', 'floatpanel')
        .style('opacity', 0);
    // has to be called after el is set
    p.el.style(p.settings.edge, pxHide(p));
    panelWidth(p.settings.width);
    if (p.settings.height) {
        panelHeight(p.settings.height);
    }
    panels[id] = p;
    /**
     * @param {?} cb
     * @return {?}
     */
    function showPanel(cb) {
        /** @type {?} */
        var endCb = fs.isF(cb) || noop;
        p.on = true;
        p.el.transition().duration(p.settings.xtnTime)
            .style(p.settings.edge, pxShow(p))
            .style('opacity', 1);
    }
    /**
     * @param {?} cb
     * @return {?}
     */
    function hidePanel(cb) {
        /** @type {?} */
        var endCb = fs.isF(cb) || noop;
        /** @type {?} */
        var endOpacity = p.settings.fade ? 0 : 1;
        p.on = false;
        p.el.transition().duration(p.settings.xtnTime)
            .style(p.settings.edge, pxHide(p))
            .style('opacity', endOpacity);
    }
    /**
     * @param {?} cb
     * @return {?}
     */
    function togglePanel(cb) {
        if (p.on) {
            hidePanel(cb);
        }
        else {
            showPanel(cb);
        }
        return p.on;
    }
    /**
     * @return {?}
     */
    function emptyPanel() {
        return p.el.text('');
    }
    /**
     * @param {?} what
     * @return {?}
     */
    function appendPanel(what) {
        return p.el.append(what);
    }
    /**
     * @param {?} w
     * @return {?}
     */
    function panelWidth(w) {
        if (w === undefined) {
            return widthVal(p);
        }
        p.el.style('width', w + 'px');
    }
    /**
     * @param {?} h
     * @return {?}
     */
    function panelHeight(h) {
        if (h === undefined) {
            return heightVal(p);
        }
        p.el.style('height', h + 'px');
    }
    /**
     * @return {?}
     */
    function panelBBox() {
        return p.el.node().getBoundingClientRect();
    }
    /**
     * @return {?}
     */
    function panelIsVisible() {
        return p.on;
    }
    /**
     * @param {?} cls
     * @param {?} bool
     * @return {?}
     */
    function classed(cls, bool) {
        return p.el.classed(cls, bool);
    }
    /**
     * @return {?}
     */
    function panelEl() {
        return p.el;
    }
    return api;
}
/**
 * @param {?} id
 * @return {?}
 */
function removePanel(id) {
    panelLayer.select('#' + id).remove();
    delete panels[id];
}
var PanelService = /** @class */ (function () {
    function PanelService(funcs, log, ts, wss) {
        this.funcs = funcs;
        this.log = log;
        this.ts = ts;
        this.wss = wss;
        fs = this.funcs;
        init();
    }
    /**
     * @param {?} id
     * @param {?} opts
     * @return {?}
     */
    PanelService.prototype.createPanel = /**
     * @param {?} id
     * @param {?} opts
     * @return {?}
     */
    function (id, opts) {
        /** @type {?} */
        var settings = Object.assign({}, defaultSettings, opts);
        if (!id) {
            this.log.warn('createPanel: no ID given');
            return null;
        }
        if (panels[id]) {
            this.log.warn('Panel with ID "' + id + '" already exists');
            return null;
        }
        if (fs.debugOn('widget')) {
            this.log.debug('creating panel:', id, settings);
        }
        return makePanel(id, settings);
    };
    /**
     * @param {?} id
     * @return {?}
     */
    PanelService.prototype.destroyPanel = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        if (panels[id]) {
            if (fs.debugOn('widget')) {
                this.log.debug('destroying panel:', id);
            }
            removePanel(id);
        }
        else {
            if (fs.debugOn('widget')) {
                this.log.debug('no panel to destroy:', id);
            }
        }
    };
    PanelService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                },] },
    ];
    /** @nocollapse */
    PanelService.ctorParameters = function () { return [
        { type: FnService },
        { type: LogService },
        { type: ThemeService },
        { type: WebSocketService }
    ]; };
    /** @nocollapse */ PanelService.ngInjectableDef = i0.defineInjectable({ factory: function PanelService_Factory() { return new PanelService(i0.inject(i1.FnService), i0.inject(i2.LogService), i0.inject(i3.ThemeService), i0.inject(i4.WebSocketService)); }, token: PanelService, providedIn: "root" });
    return PanelService;
}());
export { PanelService };
if (false) {
    /** @type {?} */
    PanelService.prototype.funcs;
    /** @type {?} */
    PanelService.prototype.log;
    /** @type {?} */
    PanelService.prototype.ts;
    /** @type {?} */
    PanelService.prototype.wss;
}
